From 93398bed1f764febf9767db8c21a6ec3bec3c6bc Mon Sep 17 00:00:00 2001
From: Thang Tran <thuutran@amperecomputing.com>
Date: Wed, 21 Aug 2024 08:58:18 +0700
Subject: [PATCH 4/5] bmc: change BMC's state to UpdateInProgress while BMC is
 in Update mode

As defined in the PDIs [1], the BMC's state is UpdateInProgress while
BMC is in Update Mode, but the phosphor-bmc-state-manager did not handle
this state.

The phosphor-software-manager is handling update actions, when users
request to update bios/bmc ..., the phosphor-software-manager creates
ActivationBlocksTransition interface [2] to block transition actions.

This commit adds monitoring to ActivationBlocksTransition interface. If
this interface is added, the BMC's state should be updated to
UpdateInProgress value. If this interface is removed, the BMC's state
should be updated to Ready.

[1] https://github.com/openbmc/phosphor-dbus-interfaces/blob/855b46f32ad385b9e26cf3a2c1e3a547e0eb8632/yaml/xyz/openbmc_project/State/BMC.interface.yaml#L57
[2] https://github.com/openbmc/phosphor-dbus-interfaces/blob/a1347418307d31a94bd21f22897aa508df938dcf/yaml/xyz/openbmc_project/Software/ActivationBlocksTransition.interface.yaml

Tested:
    1. Check the BMC's state.
       $busctl get-property xyz.openbmc_project.State.BMC \
               /xyz/openbmc_project/state/bmc0 \
               xyz.openbmc_project.State.BMC CurrentBMCState
    2. The BMC's state is "Ready".
    3. Request to flash new BIOS via Redfish/WebUI.
    4. Check the BMC's state as step 1.
    5. The BMC's state is "UpdateInProgress".
    6. Wait until updating finished.
    7. Check the BMC's state as step 1.
    8. The BMC's state is "Ready".

Change-Id: I51aa143523ec722a27eb41a1f76c3614ab6e4914
Signed-off-by: Thang Tran <thuutran@amperecomputing.com>
---
 bmc_state_manager.cpp | 42 ++++++++++++++++++++++++++++++++++++++++++
 bmc_state_manager.hpp | 38 +++++++++++++++++++++++++++++++++++++-
 2 files changed, 79 insertions(+), 1 deletion(-)

diff --git a/bmc_state_manager.cpp b/bmc_state_manager.cpp
index c891376..6a26b24 100644
--- a/bmc_state_manager.cpp
+++ b/bmc_state_manager.cpp
@@ -5,14 +5,19 @@
 
 #include <gpiod.h>
 
+#include <boost/container/flat_map.hpp>
 #include <phosphor-logging/elog-errors.hpp>
 #include <phosphor-logging/lg2.hpp>
 #include <sdbusplus/exception.hpp>
 
+#include <algorithm>
 #include <filesystem>
 #include <fstream>
 #include <iostream>
 
+using variantValues =
+    std::variant<bool, uint, uint64_t, std::string, std::vector<std::string>>;
+
 namespace phosphor
 {
 namespace state
@@ -41,6 +46,8 @@ constexpr auto SYSTEMD_SERVICE = "org.freedesktop.systemd1";
 constexpr auto SYSTEMD_OBJ_PATH = "/org/freedesktop/systemd1";
 constexpr auto SYSTEMD_INTERFACE = "org.freedesktop.systemd1.Manager";
 constexpr auto SYSTEMD_PRP_INTERFACE = "org.freedesktop.DBus.Properties";
+constexpr auto ACT_BLOCK_TRANS_INTERFACE =
+    "xyz.openbmc_project.Software.ActivationBlocksTransition";
 
 std::string BMC::getUnitState(const std::string& unitToCheck)
 {
@@ -114,6 +121,41 @@ void BMC::discoverInitialState()
     return;
 }
 
+void BMC::firmwareIsUpdating(sdbusplus::message_t& msg)
+{
+    sdbusplus::message::object_path objPath;
+    boost::container::flat_map<
+        std::string, boost::container::flat_map<std::string, variantValues>>
+        interfaces;
+    msg.read(objPath, interfaces);
+
+    /*
+     * Update the BMC state to UpdateInProgress if ActivationBlocksTransition
+     * interface is added.
+     */
+    if (interfaces.find(ACT_BLOCK_TRANS_INTERFACE) != interfaces.end())
+    {
+        this->currentBMCState(BMCState::UpdateInProgress);
+    }
+}
+
+void BMC::firmwareIsUpdated(sdbusplus::message_t& msg)
+{
+    sdbusplus::message::object_path objPath;
+    std::vector<std::string> interfaces;
+    msg.read(objPath, interfaces);
+
+    /*
+     * Update the BMC state to Ready if ActivationBlocksTransition interface is
+     * removed.
+     */
+    if (std::find(interfaces.begin(), interfaces.end(),
+                  ACT_BLOCK_TRANS_INTERFACE) != interfaces.end())
+    {
+        this->discoverInitialState();
+    }
+}
+
 void BMC::executeTransition(const Transition tranReq)
 {
     // HardReboot does not shutdown any services and immediately transitions
diff --git a/bmc_state_manager.hpp b/bmc_state_manager.hpp
index 50d25f1..b1a59f4 100644
--- a/bmc_state_manager.hpp
+++ b/bmc_state_manager.hpp
@@ -66,7 +66,19 @@ class BMC : public BMCInherit
                         timeSyncSignal.reset();
                     }
                 }
-            }))
+            })),
+
+        fwUpdatingSignal(
+            std::make_unique<decltype(fwUpdatingSignal)::element_type>(
+                bus,
+                sdbusRule::interfacesAdded("/xyz/openbmc_project/software"),
+                [this](sdbusplus::message_t& m) { firmwareIsUpdating(m); })),
+
+        fwUpdatedSignal(
+            std::make_unique<decltype(fwUpdatedSignal)::element_type>(
+                bus,
+                sdbusRule::interfacesRemoved("/xyz/openbmc_project/software"),
+                [this](sdbusplus::message_t& m) { firmwareIsUpdated(m); }))
     {
         utils::subscribeToSystemdSignals(bus);
         discoverInitialState();
@@ -121,6 +133,24 @@ class BMC : public BMCInherit
      */
     int bmcStateChange(sdbusplus::message_t& msg);
 
+    /** @brief Callback function when BMC in update mode
+     *
+     * Update the BMC's state to UpdateInProgress while BMC in update mode
+     *
+     * @param[in]  msg       - Data associated with subscribed signal
+     *
+     */
+    void firmwareIsUpdating(sdbusplus::message_t& msg);
+
+    /** @brief Callback function on BMC updated firmware
+     *
+     * Update the BMC's state to Ready when update is done
+     *
+     * @param[in]  msg       - Data associated with subscribed signal
+     *
+     */
+    void firmwareIsUpdated(sdbusplus::message_t& msg);
+
     /** @brief Persistent sdbusplus DBus bus connection. **/
     sdbusplus::bus_t& bus;
 
@@ -130,6 +160,12 @@ class BMC : public BMCInherit
     /** @brief Used to subscribe to timesync **/
     std::unique_ptr<sdbusplus::bus::match_t> timeSyncSignal;
 
+    /** @brief Used to subscribe to BMC is in update mode **/
+    std::unique_ptr<sdbusplus::bus::match_t> fwUpdatingSignal;
+
+    /** @brief Used to subscribe to BMC updated firmware **/
+    std::unique_ptr<sdbusplus::bus::match_t> fwUpdatedSignal;
+
     /**
      * @brief discover the last reboot cause of the bmc
      **/
-- 
2.34.1

